packages/subscription/src/stream.ts

```diff
@@ -28,16 +30,20 @@ export interface StringBufferStream<TReturn = undefined> extends Stream<string,
 export function createStream<T, TReturn = undefined>(callback: Callback<T, TReturn>): Stream<T, TReturn> {
   let iterable: ToOperationIterator<T, TReturn> = (task) => createOperationIterator(task, callback);
 
-  let subscribable = {
-    filter(predicate: (value: T) => boolean): Stream<T, TReturn> {
-      return createStream((publish) => {
-        return subscribable.forEach((value) => function*() {
-          if(predicate(value)) {
-            publish(value);
-          }
-        });
+  function filter<R extends T>(predicate: (value: T) => value is R): Stream<T, TReturn>
+  function filter(predicate: (value: T) => boolean): Stream<T, TReturn>
+  function filter(predicate: (value: T) => boolean): Stream<T, TReturn> {
```

Comment:
Would be really nice not to have to write out this ridiculous type, but I couldn't find a way to do that.